File: src/Shared/Logistics.Shared.Identity/Policies/Permissions.cs
using System.Reflection;

namespace Logistics.Shared.Identity.Policies;

public static class Permission
{
    private static readonly Dictionary<string, IEnumerable<string>> Modules;

    static Permission()
    {
        Modules = typeof(Permission)
            .GetNestedTypes(BindingFlags.Public | BindingFlags.Static)
            .Where(type => type is { IsClass: true, IsSealed: true, IsAbstract: true })
            .ToDictionary(
                type => type.Name,
                type => type.GetFields(BindingFlags.Public | BindingFlags.Static)
                    .Where(field => field is { IsLiteral: true, IsInitOnly: false })
                    .Select(field => (string)field.GetValue(null)!)
            );
    }

    public static class AppRole
    {
        public const string View = $"{nameof(Permission)}.{nameof(AppRole)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(AppRole)}.Manage";
    }

    public static class Employee
    {
        public const string View = $"{nameof(Permission)}.{nameof(Employee)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Employee)}.Manage";
    }

    public static class Customer
    {
        public const string View = $"{nameof(Permission)}.{nameof(Customer)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Customer)}.Manage";
    }

    public static class Driver
    {
        public const string View = $"{nameof(Permission)}.{nameof(Driver)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Driver)}.Manage";
    }

    public static class Notification
    {
        public const string View = $"{nameof(Permission)}.{nameof(Notification)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Notification)}.Manage";
    }

    public static class Load
    {
        public const string View = $"{nameof(Permission)}.{nameof(Load)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Load)}.Manage";
    }

    public static class Stat
    {
        public const string View = $"{nameof(Permission)}.{nameof(Stat)}.View";
    }

    public static class Tenant
    {
        public const string View = $"{nameof(Permission)}.{nameof(Tenant)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Tenant)}.Manage";
    }

    public static class TenantRole
    {
        public const string View = $"{nameof(Permission)}.{nameof(TenantRole)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(TenantRole)}.Manage";
    }

    public static class Truck
    {
        public const string View = $"{nameof(Permission)}.{nameof(Truck)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Truck)}.Manage";
    }

    public static class User
    {
        public const string View = $"{nameof(Permission)}.{nameof(User)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(User)}.Manage";
    }

    public static class Payment
    {
        public const string View = $"{nameof(Permission)}.{nameof(Payment)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Payment)}.Manage";
    }

    public static class Invoice
    {
        public const string View = $"{nameof(Permission)}.{nameof(Invoice)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Invoice)}.Manage";
    }

    public static class Payroll
    {
        public const string View = $"{nameof(Permission)}.{nameof(Payroll)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Payroll)}.Manage";
        public const string Approve = $"{nameof(Permission)}.{nameof(Payroll)}.Approve";
    }

    public static class Eld
    {
        public const string View = $"{nameof(Permission)}.{nameof(Eld)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Eld)}.Manage";
        public const string Sync = $"{nameof(Permission)}.{nameof(Eld)}.Sync";
    }

    public static class Message
    {
        public const string View = $"{nameof(Permission)}.{nameof(Message)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Message)}.Manage";
    }

    public static class Invitation
    {
        public const string View = $"{nameof(Permission)}.{nameof(Invitation)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Invitation)}.Manage";
    }

    public static class Portal
    {
        public const string Access = $"{nameof(Permission)}.{nameof(Portal)}.Access";
        public const string ViewLoads = $"{nameof(Permission)}.{nameof(Portal)}.ViewLoads";
        public const string ViewInvoices = $"{nameof(Permission)}.{nameof(Portal)}.ViewInvoices";
        public const string ViewDocuments = $"{nameof(Permission)}.{nameof(Portal)}.ViewDocuments";
    }

    public static class Expense
    {
        public const string View = $"{nameof(Permission)}.{nameof(Expense)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Expense)}.Manage";
        public const string Approve = $"{nameof(Permission)}.{nameof(Expense)}.Approve";
    }

    public static class LoadBoard
    {
        public const string View = $"{nameof(Permission)}.{nameof(LoadBoard)}.View";
        public const string Search = $"{nameof(Permission)}.{nameof(LoadBoard)}.Search";
        public const string Book = $"{nameof(Permission)}.{nameof(LoadBoard)}.Book";
        public const string Post = $"{nameof(Permission)}.{nameof(LoadBoard)}.Post";
        public const string Manage = $"{nameof(Permission)}.{nameof(LoadBoard)}.Manage";
    }

    public static class BlogPost
    {
        public const string View = $"{nameof(Permission)}.{nameof(BlogPost)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(BlogPost)}.Manage";
    }

    public static class Dvir
    {
        public const string View = $"{nameof(Permission)}.{nameof(Dvir)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Dvir)}.Manage";
        public const string Review = $"{nameof(Permission)}.{nameof(Dvir)}.Review";
    }

    public static class Safety
    {
        public const string View = $"{nameof(Permission)}.{nameof(Safety)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Safety)}.Manage";
    }

    public static class Maintenance
    {
        public const string View = $"{nameof(Permission)}.{nameof(Maintenance)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Maintenance)}.Manage";
    }

    public static class Certification
    {
        public const string View = $"{nameof(Permission)}.{nameof(Certification)}.View";
        public const string Manage = $"{nameof(Permission)}.{nameof(Certification)}.Manage";
        public const string Verify = $"{nameof(Permission)}.{nameof(Certification)}.Verify";
    }

    public static IEnumerable<string> GetAll()
    {
        var list = new List<string>();

        foreach (var module in Modules.Keys)
        {
            list.AddRange(GeneratePermissions(module));
        }
        return list;
    }

    public static IEnumerable<string> GeneratePermissions(string module)
    {
        return Modules[module];
    }
}

File: src/Client/Logistics.Angular/projects/shared/src/lib/models/permissions.ts
export const Permission = {
  AppRole: {
    View: "Permission.AppRole.View",
    Manage: "Permission.AppRole.Manage",
  },
  Employee: {
    View: "Permission.Employee.View",
    Manage: "Permission.Employee.Manage",
  },
  Load: {
    View: "Permission.Load.View",
    Manage: "Permission.Load.Manage",
  },
  Notification: {
    View: "Permission.Notification.View",
    Manage: "Permission.Notification.Manage",
  },
  Stat: {
    View: "Permission.Stat.View",
  },
  Tenant: {
    View: "Permission.Tenant.View",
    Manage: "Permission.Tenant.Manage",
  },
  TenantRole: {
    View: "Permission.TenantRole.View",
    Manage: "Permission.TenantRole.Manage",
  },
  Truck: {
    View: "Permission.Truck.View",
    Manage: "Permission.Truck.Manage",
  },
  User: {
    View: "Permission.User.View",
    Manage: "Permission.User.Manage",
  },
  Customer: {
    View: "Permission.Customer.View",
    Manage: "Permission.Customer.Manage",
  },
  Payment: {
    View: "Permission.Payment.View",
    Manage: "Permission.Payment.Manage",
  },
  Invoice: {
    View: "Permission.Invoice.View",
    Manage: "Permission.Invoice.Manage",
  },
  Tax: {
    View: "Permission.Tax.View",
    Manage: "Permission.Tax.Manage",
  },
  Accounting: {
    View: "Permission.Accounting.View",
    Manage: "Permission.Accounting.Manage",
  },
  ApiKey: {
    View: "Permission.ApiKey.View",
    Manage: "Permission.ApiKey.Manage",
  },
  Payroll: {
    View: "Permission.Payroll.View",
    Manage: "Permission.Payroll.Manage",
  },
  Eld: {
    View: "Permission.Eld.View",
    Manage: "Permission.Eld.Manage",
    Sync: "Permission.Eld.Sync",
  },
  Message: {
    View: "Permission.Message.View",
    Manage: "Permission.Message.Manage",
  },
  Invitation: {
    View: "Permission.Invitation.View",
    Manage: "Permission.Invitation.Manage",
  },
  Portal: {
    Access: "Permission.Portal.Access",
    ViewLoads: "Permission.Portal.ViewLoads",
    ViewInvoices: "Permission.Portal.ViewInvoices",
    ViewDocuments: "Permission.Portal.ViewDocuments",
  },
  Expense: {
    View: "Permission.Expense.View",
    Manage: "Permission.Expense.Manage",
  },
  LoadBoard: {
    View: "Permission.LoadBoard.View",
    Search: "Permission.LoadBoard.Search",
    Book: "Permission.LoadBoard.Book",
    Post: "Permission.LoadBoard.Post",
    Manage: "Permission.LoadBoard.Manage",
  },
  FuelCard: {
    View: "Permission.FuelCard.View",
    Manage: "Permission.FuelCard.Manage",
  },
  Dispatch: {
    View: "Permission.Dispatch.View",
    Manage: "Permission.Dispatch.Manage",
  },
  BlogPost: {
    View: "Permission.BlogPost.View",
    Manage: "Permission.BlogPost.Manage",
  },
  IftaRate: {
    View: "Permission.IftaRate.View",
    Manage: "Permission.IftaRate.Manage",
  },
  Document: {
    View: "Permission.Document.View",
    Manage: "Permission.Document.Manage",
    Review: "Permission.Document.Review",
  },
  Dvir: {
    View: "Permission.Dvir.View",
    Manage: "Permission.Dvir.Manage",
    Review: "Permission.Dvir.Review",
  },
  Safety: {
    View: "Permission.Safety.View",
    Manage: "Permission.Safety.Manage",
  },
  Maintenance: {
    View: "Permission.Maintenance.View",
    Manage: "Permission.Maintenance.Manage",
  },
  Accident: {
    View: "Permission.Accident.View",
    Manage: "Permission.Accident.Manage",
    Review: "Permission.Accident.Review",
  },
  Container: {
    View: "Permission.Container.View",
    Manage: "Permission.Container.Manage",
  },
  Copilot: {
    View: "Permission.Copilot.View",
    Manage: "Permission.Copilot.Manage",
  },
  Terminal: {
    View: "Permission.Terminal.View",
    Manage: "Permission.Terminal.Manage",
  },
  Negotiation: {
    View: "Permission.Negotiation.View",
    Manage: "Permission.Negotiation.Manage",
  },
  ProductLicense: {
    View: "Permission.ProductLicense.View",
    Manage: "Permission.ProductLicense.Manage",
  },
} as const;

/**
 * Type representing all possible permission values.
 * Extracted from the Permission object for strong typing.
 */
export type PermissionValue = {
  [K in keyof typeof Permission]: (typeof Permission)[K][keyof (typeof Permission)[K]];
}[keyof typeof Permission];
